tensordict/_reductions.py

```diff
@@ -0,0 +1,152 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+from __future__ import annotations
+
+import copyreg
+from multiprocessing.reduction import ForkingPickler
+
+import torch
+from tensordict._lazy import LazyStackedTensorDict
+from tensordict._td import TensorDict
+
+from tensordict.tensorclass import NonTensorData
+from tensordict.utils import _STRDTYPE2DTYPE
+
+CLS_MAP = {
+    "TensorDict": TensorDict,
+    "LazyStackedTensorDict": LazyStackedTensorDict,
+}
+
+
+def _rebuild_tensordict_files(flat_key_values, metadata_dict, is_shared: bool = False):
+    def from_metadata(metadata=metadata_dict, prefix=None):
+        non_tensor = metadata.pop("non_tensors")
+        leaves = metadata.pop("leaves")
+        cls = metadata.pop("cls")
+        cls_metadata = metadata.pop("cls_metadata")
+        is_locked = cls_metadata.pop("is_locked", False)
+
+        d = {
+            key: NonTensorData(data, batch_size=batch_size)
+            for (key, (data, batch_size)) in non_tensor.items()
+        }
+        for key, _ in leaves.items():
+            total_key = (key,) if prefix is None else prefix + (key,)
+            if total_key[-1].startswith("<NJT>"):
+                nested_values = flat_key_values[total_key]
+                continue
+            if total_key[-1].startswith("<NJT_OFFSETS"):
+                offsets = flat_key_values[total_key]
+                key = key.replace("<NJT_OFFSETS>", "")
+                value = torch.nested.nested_tensor_from_jagged(nested_values, offsets)
```

Comment:
nit: eventually it'll likely be useful to support NJT's `lengths` as well, which is an optional set of metadata present for non-contiguous NJTs

Reply:
Do you have an example of such a thing?

Reply:
sure thing, here's a test case demonstrating construction with lengths: https://github.com/pytorch/pytorch/blob/d52684e9a8b95f8b0d06f0bf08e6a39846cb3ae6/test/test_nestedtensor.py#L4242

When both `offsets` (shape `B + 1` including the end offset) and `lengths` (shape `B`) are specified, the idea is that `offsets` point to the beginning of each batch item sequence and `lengths` indicates the length of each. This allows for representing NJTs that are "non-contiguous with holes". If only `offsets` is specified, the assumption is that the entire length between offsets is part of each batch item, which of course doesn't allow for such holes.